### PR TITLE
primitives: Add initial proof-of-work funcs.

### PR DESCRIPTION
**Pull request by davecgh**
~**This requires #2787**.~

This introduces a primitives package that currently only contains functions related to proof of work along with comprehensive tests, benchmarks, full documentation and a skeleton `README.md`.

These functions are all the semantic equivalents of their respective functions in `blockchain/standalone` updated to use, accept, and return the new uint256 type instead of stdlib big integers.

Specifically, the following functions are updated and added along with their associated tests and benchmarks:

- `DiffBitsToUint256` (equiv to `standalone.CompactToBig`)
- `Uint256ToDiffBits` (equiv to `standalone.BigToCompact`)
- `CalcWork` (equiv to `standalone.CalcWork`)
- `HashToUint256`  (equiv to `standalone.HashToBig`)
- `CheckProofOfWorkRange`  (equiv to `standalone.CheckProofOfWorkRange`)
- `CheckProofOfWork`  (equiv to `standalone.CheckProofOfWork`)

It is split into multiple commits to ease the review process and each commit further describes the relevant changes.

A few key points:

The primary motivation for renaming the functions to specifically call out difficulty bits instead of the shorter "Compact" is to prevent future ambiguity with Satoshi-style variable length compact integers that will ultimately also be available from the package.

For legacy reasons, even though the difficulty bits encoding is only used in Decred to encode unsigned 256-bit integers which represent difficulty targets, the encoding is capable of representing negative numbers as well as numbers much larger than the maximum value of an unsigned 256-bit integer.  Therefore, in order to enable proper error detection and stay consistent with legacy code, `DiffBitsToUint256` returns additional flags to indicate the aforementioned conditions.

The original calculation that `CalcWork` performed (and therefore is used in consensus and must stay the same) involves a dividend of 2^256 which is not directly representable by a uint256, so this implementation retains the same semantics by transforming the calculation as described in detail by the comments in the code.

## Benchmarks

The following is a comparison between the existing implementations in the blockchain/standalone module with big integers (old) and the new implementation (new) averaging 10 runs each:

Name              |  Old Time/Op  |  New Time/Op  | Delta
------------------|---------------|---------------|--------
DiffBitsToUint256 | 158ns ± 1%    | 4ns ± 1%      | -97.35%
Uint256ToDiffBits | 151ns ± 2%    | 8ns ± 4%      | -94.82%
CalcWork          | 638ns ± 1%    | 59ns ± 1%     | -90.78%
HashToUint256     | 127ns ± 3%    | 3ns ± 1%      | -97.68%
CheckProofOfWork  | 314ns ± 1%    | 43ns ± 0%     | -86.42%

Name              | Old Allocs/Op | New Allocs/Op | Delta
------------------|---------------|---------------|---------
DiffBitsToUint256 |   3.00 ± 0%   |    0.00       | -100.00%
Uint256ToDiffBits |   2.00 ± 0%   | 0.00          | -100.00%
CalcWork          |   7.00 ± 0%   | 0.00          | -100.00%
HashToUint256     |   2.00 ± 0%   | 0.00          | -100.00%
CheckProofOfWork  |   5.00 ± 0%   | 0.00          | -100.00%

This is work towards #2786.

### CODE REVIEW

File: internal/staging/primitives/pow.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	// to fit into the available 23-bits, so divide the number by 256 and
+	// increment the exponent accordingly.
+	if mantissa&0x00800000 != 0 {
+		mantissa >>= 8
+		exponent++
+	}
+
+	// Pack the exponent and mantissa into an unsigned 32-bit int and return it.
+	//
+	// Note that the sign bit is always unset here because uint256s can never be
```
Comment (davecgh) [inline internal/staging/primitives/pow.go:128]: This is leftover from before the flag to force it to be negative was added.  It needs to be updated to reflect reality.

**Review by davecgh — COMMENTED** [review]
Commit title typo `rimitives: Add check proof of work. `.

File: internal/staging/primitives/pow.go (hunk 1 above)
Comment (davecgh) [reply internal/staging/primitives/pow.go]: Updated.

**Comment by davecgh** [comment]
> Commit title typo `rimitives: Add check proof of work. `.

Corrected.

**Review by rstaudt2 — COMMENTED** [review]
Looks good to me.  I confirmed that the behavior is the same as the corresponding functions in `blockchain/standalone`.  I just had a few minor comments inline.

File: internal/staging/primitives/pow.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+// negative value and/or overflows a uint256 to enable proper error detection
+// and stay consistent with legacy code.
+func DiffBitsToUint256(bits uint32) (n uint256.Uint256, isNegative bool, overflows bool) {
+	// Extract the mantissa, sign bit, and exponent.
+	mantissa := bits & 0x007fffff
+	isSignBitSet := bits&0x00800000 != 0
+	exponent := bits >> 24
+
+	// Nothing to do when the mantissa is zero as any multiple of it will
+	// necessarily also be 0 and therefore it can never be negative of overflow.
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow.go:49]: Fixed.

I also reworded the comments in the following sections a bit to clarify things better since they were just copied from old code and I noticed they probably weren't super clear without already knowing that history because it talks about bytes which are no longer used.

File: internal/staging/primitives/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+[![Doc](https://img.shields.io/badge/doc-reference-blue.svg)](https://pkg.go.dev/github.com/decred/dcrd/internal/staging/primitives)
+
+## Package and Module Status
+
+This package is currently a work in progress in the context of a larger refactor
+and thus does not yet provide most of the things that are ultimately planned.
+See https://github.com/decred/dcrd/issues/2786 for further details.
+
+The intention is to create a containing `primitives` module that will be kept at
+an experimental module version ("v0") until everything is stablized to avoid
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/README.md:15]: Fixed.

File: internal/staging/primitives/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+This package ultimately aims to provide core data structures and functions for
+working with several aspects of Decred consensus.
+
+The provided functions fall into the following categories:
+
+- Proof-of-work
+  - Converting to and from the target difficulty bits representation
+  - Calculating work values based on the target difficulty bits
+  - Checking a block hash satisfies a target difficulty and that target
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/README.md:28]: Added.

File: internal/staging/primitives/pow_test.go [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+	if len(b) > 32 {
+		panic("hex in source file overflows mod 2^256: " + s)
+	}
+	return new(uint256.Uint256).SetByteSlice(b)
+}
+
+// TestDiffBitsToUint256 ensures converting from the compact representation used
+// for target difficulties to unsigned 256-bit integers produces the correct
+// results.
+func TestDiffBitsToUint256(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow_test.go:46]: Done.  Although, I think there are currently so few tests that the overhead of running them in parallel costs more than it saves, but there will end up being a lot more tests in the future, so it should eventually up being a savings.

```
Without: github.com/decred/dcrd/internal/staging/primitives      0.167s
With:    github.com/decred/dcrd/internal/staging/primitives      0.193s
```

**Review by rstaudt2 — APPROVED** [review]

**Review by JoeGruffins — COMMENTED** [review]
<details closed>
<summary>benchmarks</summary>

```
$ go test -bench .
goos: linux
goarch: amd64
pkg: github.com/decred/dcrd/internal/staging/primitives
cpu: AMD Ryzen 9 3900XT 12-Core Processor           
BenchmarkDiffBitsToUint256-24           332246042                3.625 ns/op           0 B/op          0 allocs/op
BenchmarkUint256ToDiffBits-24           220453808                5.322 ns/op           0 B/op          0 allocs/op
BenchmarkCalcWork-24                    28234624                41.63 ns/op            0 B/op          0 allocs/op
BenchmarkHashToUint256-24               500613214                2.060 ns/op           0 B/op          0 allocs/op
BenchmarkCheckProofOfWork-24            39382293                29.22 ns/op            0 B/op          0 allocs/op
PASS
ok      github.com/decred/dcrd/internal/staging/primitives      8.405s
```
</details>

File: internal/staging/primitives/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+- Be extremely careful when making any changes to avoid breaking consensus
+  - This often means existing code can't be changed without adding an internal
+    flag to control behavior and introducing a new method with the new behavior
+- Minimize the number of allocations as much as possible
+- Opt for data structures that improve cache locality
+- Keep a strong focus on providing efficient code
+- Avoid external dependencies
+  - Consensus code requires much stronger guarantees than typical code and
+    consequently code that is not specifically designed with such rigid
+    constraints will almost invariably inevitably break things in subtle ways
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/README.md:46]: They don't mean the same thing.  I'll reword it to "almost always eventually" if you find the clearer.

File: internal/staging/primitives/pow.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+func uint256ToDiffBits(n *uint256.Uint256, isNegative bool) uint32 {
+	// No need to do any work if it's zero.
+	if n.IsZero() {
+		return 0
+	}
+
+	// Since the base for the exponent is 256, the exponent can be treated as
+	// the number of bytes it takes to represent the value.  So, shift the
+	// number right or left accordingly.  This is equivalent to:
+	// mantissa /= 256^(exponent-3)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow.go]: Fair enough.  Updated.  The mantissa was understood to start as n there, but it's good to be more explicit.

File: internal/staging/primitives/pow.go [hunk 8]
```diff
@@ -141,3 +141,50 @@ func Uint256ToDiffBits(n *uint256.Uint256) uint32 {
...
+
+	// The goal is to calculate 2^256 / (diff+1), where diff > 0 using a
+	// fixed-precision uint256.
+	//
+	// Since 2^256 can't be represented by a uint256, the calc is performed as
+	// follows:
+	//
+	// Notice:
+	//    work = (2^256 / (diff+1))
+	// => work = ((2^256-diff-1) / (diff+1))+1
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow.go:177]: Yes, that's works, although I think it's perhaps a little easier to see by just adding zero in the form of `-1 + 1` so you don't have to mess with the left hand side of the equation:

```
=> (2^256 / (diff+1)) - 1 + 1
=> (2^256 / (diff+1)) - (diff+1)/(diff+1) + 1
=> ((2^256-diff-1) / (diff+1)) + 1
```

File: internal/staging/primitives/pow_test.go [hunk 9]
```diff
@@ -209,3 +209,49 @@ func TestUint256ToDiffBits(t *testing.T) {
...
+		input uint32 // target difficulty bits to test
+		want  string // expected uint256
+	}{{
+		name:  "mainnet block 1",
+		input: 0x1b01ffff,
+		want:  "0000000000000000000000000000000000000000000000000000800040002000",
+	}, {
+		name:  "mainnet block 288",
+		input: 0x1b01330e,
+		want:  "0000000000000000000000000000000000000000000000000000d56f2dcbe105",
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow_test.go:231]: `chainwork` is the **cumulative** work up to block 288.  This is testing the work value for **only** block 288.

File: internal/staging/primitives/pow_test.go [hunk 10]
```diff
@@ -209,3 +209,49 @@ func TestUint256ToDiffBits(t *testing.T) {
...
+// difficulty produces the correct results.
+func TestCalcWork(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name  string // test description
+		input uint32 // target difficulty bits to test
+		want  string // expected uint256
+	}{{
+		name:  "mainnet block 1",
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow_test.go:225]: The genesis block is valid by definition and therefore was never mined, so it doesn't apply.

File: internal/staging/primitives/pow.go [hunk 11]
```diff
@@ -198,3 +200,62 @@ func HashToUint256(hash *chainhash.Hash) uint256.Uint256 {
...
+	// than the maximum value that can be represented by a uint256.
+	target, isNegative, overflows := DiffBitsToUint256(difficultyBits)
+	if isNegative {
+		str := fmt.Sprintf("target difficulty bits %08x is a negative value",
+			difficultyBits)
+		return uint256.Uint256{}, ruleError(ErrUnexpectedDifficulty, str)
+	}
+	if overflows {
+		str := fmt.Sprintf("target difficulty bits %08x is higher than the "+
+			"max limit %x", difficultyBits, powLimit)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/staging/primitives/pow.go]: The max allowed value is the pow limit, but the encoded value could be much, much higher than a uint256, correct.

File: internal/staging/primitives/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply internal/staging/primitives/README.md]: yes. ty.

File: internal/staging/primitives/pow.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply internal/staging/primitives/pow.go:177]: Thank you.

File: internal/staging/primitives/pow_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (JoeGruffins) [reply internal/staging/primitives/pow_test.go:225]: I see.

**Review by JoeGruffins — APPROVED** [review]

File: internal/staging/primitives/pow.go [hunk 12]
```diff
@@ -188,3 +189,12 @@ func CalcWork(diffBits uint32) uint256.Uint256 {
 	divisor := new(uint256.Uint256).SetUint64(1).Add(&diff)
 	return *diff.Not().Div(divisor).AddUint64(1)
 }
+
+// HashToUint256 converts the provided hash to an unsigned 256-bit integer that
+// can be used to perform math comparisons.
+func HashToUint256(hash *chainhash.Hash) uint256.Uint256 {
+	// Hashes are a stream of bytes that do not have any inherit endianness to
```
Comment (davecgh) [inline internal/staging/primitives/pow.go:198]:
```suggestion
	// Hashes are a stream of bytes that do not have any inherent endianness to
```

Comment (davecgh) [reply internal/staging/primitives/pow.go]: Fixed.
